Support unobfuscated versions + bug fixes & improvements

SizableShrimp opened the pull request:
Snowblower as it currently stands has a number of deficiencies which this PR attempts to address.

Main features:
* **Raise minimum Java to 25.**
* **Support unobfuscated versions**, like `26.1-snapshot-1`. This is also set up to work automatically with future versions, like `26.1-pre-1` or `26.1-rc-1`.
  * Additionally, Minecraft versions are now parsed using a custom class rather than depending on SRGUtils. The main benefit of this system is allowing *any* future Minecraft version format to be parsed even if it is not explicitly accounted for; it just may be parsed as a "special" version and need to be explicitly included using a branch config.
  * All previously-released `_unobfuscated` version variants are also supported, but every unobfuscated version except `1.21.11_unobfuscated` require explicit inclusion using a branch config. (The complete list of versions with unobfuscated variants is `25w45a`, `25w46a`, `1.21.11-pre1`, `1.21.11-pre2`, `1.21.11-pre3`, `1.21.11-pre4`, `1.21.11-pre5`, `1.21.11-rc1`, `1.21.11-rc2`, `1.21.11-rc3`, and `1.21.11`.)
    * To support cleaner diffs when generating versions between `1.21.11` and earlier <-> `26.1` and later, `1.21.11_unobfuscated` is included in the version list by default. Since this version includes names for EVERY local variable and parameter in the game that were previously autogenerated, the `1.21.11` <-> `1.21.11_unobfuscated` diff is **extremely large** (see picture below). Unfortunately, there isn't really any way around this. It means that tracking down changes across this `1.21.11` <-> `26.1` boundary will become a lot more annoying when using git blame. But hey, at least we have names for everything now!
    ![Change summary for version diff from 1.21.11 to 1.21.11_unobfuscated](https://i.imgur.com/6Ak1Lpz.png)
* **Fix bugs and general improvements around when to start over and when to push to remote.** There was a bug with the detection logic for whether the branch is considered "up-to-date" when determining whether to push old commits, leading to the current version of Snowblower *re-pushing every commit* on startup. Fixing this alone speeds up runtime by a lot.
  * Other improvements have been added around the `--start-over-if-required` flag, making it smarter for detecting more complicated incompatible changes and starting over the branch where appropriate. (This part needs more testing to be honest, but I don't feel like writing a test harness at the moment.)
* **Improve logging** to both log more things (e.g., computed start and end version, whether branch was checked out from remote, started over, etc.) and, when a GitHub Actions environment is detected, group log lines together (can also manually enable/disable using `--github-actions [enabled]` flag).
  * Now uses SLF4J's Mapped Diagnostic Context (MDC) system to set a Minecraft version property that is prepended to all log lines when generating a specific Minecraft version. This makes it easy to see exactly which log lines are associated with which versions.
* **Discover and download all artifacts ahead-of-time** using multiple threads (includes all mappings, client/server JARs, and libraries) to speed up runtime and minimize downloads during the middle of generation. This supersedes #11.
* **Overhaul the README** to provide more background on what the project is, what it can be used for, how to use it, and where the name "Snowblower" came from (with a history lesson)!

Closes #8.
Supersedes #11.

SizableShrimp committed:
First attempt at unobfuscated decompile

Right now, the code doesn't work because the new 26.1 snapshot also brought with it Java 25, and MergeTool uses an outdated version of ASM.

SizableShrimp committed:
Support 26.1-snapshot-1 and unobfuscated jars

Use ProcessMinecraftJar task from InstallerTools to unify merging and remapping code
Fix bug where old commits would be pushed even if the --push flag was not set
Fix bug where all commits would be pushed again if the branch was up-to-date with the remote
When generating 1.21.11, Snowblower will now additionally generate 1.21.11_unobfuscated in addition, to support cleaner diffs for unobfuscated versions after 1.21.11 (e.g., 26.1-snapshot-1).

SizableShrimp committed:
Improve logging

SizableShrimp committed:
Update version formats based on slicedlime's infographic

SizableShrimp committed:
Overhaul --start-over-if-required checks and support all unobfuscated versions

All unobfuscated versions except for 1.21.11_unobfuscated are excluded by default; they can be included by explicitly adding them to the included versions list.
Improved detection for cases when starting over the branch is necessary (rather than forcing an error on the user or starting over too often).
* Rather than restarting every time a new version of Snowblower is released, now a manually incremented version ID is used, so branches are only restarted when actually necessary.
* Snowblower will now detect if the latest committed version on a branch is newer than the target and skip generating.
* Snowblower will accurately report if the latest version either doesn't exist, is older than the start version, or filtered out by the current branch configuration (or start over with --start-over-if-required).
Improved error reporting to use Logger#error in more cases, rather than throwing an exception.
Added more logging during generation/setup so that the computed configuration is more transparent to users.

SizableShrimp committed:
Clarify that all pushes are force pushes in logging

SizableShrimp committed:
Support grouping log lines in GitHub Actions environment with `--github-actions` flag

SizableShrimp committed:
Default GitHub Actions flag to value of `GITHUB_ACTIONS` env var

SizableShrimp committed:
wip

SizableShrimp committed:
Fix support for pre-1.18 versions

Also fix general bugs with JAR file iteration
Update InstallerTools to a PR build to fix support for pre-1.16 versions

SizableShrimp committed:
Fix JAD-style var names for obfuscated versions

Overhauled partial cache checking to be smarter; the partial cache for the decompiled jar will now be skipped if any individual task is outdated in the partial cache.
Extracted decompilation from Generator to DecompileTask

SizableShrimp committed:
Fix dist annotations for older versions

Use MergeTool for obfuscated game versions to respect dist annotations on class members
Changed unobfuscated game versions to add dist annotations with ProcessMinecraftJar instead of the OnlyInPlugin provided by neoforged's VineflowerPlugins, since VF's plugin system currently has bugs (and the plugin itself does not respect dist annotations on nested classes)

SizableShrimp committed:
Update InstallerTools to published version

SizableShrimp committed:
Overhaul README and add April Fools' Day support

SizableShrimp committed:
Add ahead-of-time multithreaded artifact discovery

Matyrobbrt commented on src/main/java/net/neoforged/snowblower/data/Version.java:
```diff
@@ -26,13 +25,19 @@ public record Version(
     }
 
+    public boolean isUnobfuscated() {
+        return FIRST_UNOBFUSCATED_RELEASE_DATE.compareTo(this.releaseTime) <= 0 || this.id.version().endsWith("_unobfuscated");
```
I don't think this is a guarantee and I'd instead compare the version against 26.1-snapshot-1. There could always be a 1.21.12 until 26.1 is released.

SizableShrimp replied:
Hmm, this poses a bit of a problem because by rewriting MinecraftVersion to use a custom class, I no longer have access to comparisons, at least not without consulting the version manifest. (And the manifest would do no better here.)

The main reason for this change is it simplifies the MinecraftVersion class a lot, and I don't need a huge if block for the snapshots that needs to be maintained to determine sorting. And it hasn't really been necessary, since the version manifest gives sorting.

While I understand your point, I think it's unlikely to come up. How often do they release new versions for the current version when they have started the next development cycle? Also, it debatable whether or not they will designate 1.21.12 (if such a version were to come out) as unobfuscated by default or not. Thus, I think it's reasonable to keep this as-is for now and reevaluate if it turns out not to be the case.

sciwhiz12 replied:
This far into the development cycle, I think Mojang is only going to release a new hotfix when there's a security issue or massive game-breaking flaw. I think we can proceed with this heuristic for now, and cross that bridge when we get to it as Shrimp says.

SizableShrimp commented on src/main/java/net/neoforged/snowblower/util/ArtifactDiscoverer.java:
```diff
@@ -0,0 +1,130 @@
+        LOGGER.info("Discovering and downloading artifacts for {} versions", versions.size());
+        GitHubActions.logStartGroup("Discovering and downloading artifacts");
+
+        try (ExecutorService executor = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors())) {
```
I guess I haven't compared them, but would an unbounded number of threads not be slowed down by high CPU contention? I can test and come back to this.

sciwhiz12 replied:
I'm going to push work on virtual threads to a future PR, since this PR is big enough already.

sciwhiz12 commented on src/main/java/net/neoforged/snowblower/data/MinecraftVersion.java:
```diff
@@ -0,0 +1,88 @@
+
+        public boolean isSpecial() {
+            return this.special;
+        }
```
Technically, you could reduce `isSpecial` to just be a check on `this == SPECIAL`.

sciwhiz12 commented on src/main/java/net/neoforged/snowblower/github/GitHubActions.java:
A thought: this could be expanded in the future to also encompass other CI systems that have equivalent functionality. [TeamCity's service messages](https://www.jetbrains.com/help/teamcity/service-messages.html#Blocks+of+Service+Messages) comes to mind in this regard. But that can be implemented in the future.

sciwhiz12 commented on src/main/java/net/neoforged/snowblower/tasks/BundlerExtractTask.java:
```diff
@@ -19,46 +21,124 @@
+                .put("server", version.downloads().get("server").sha1());
+    }
+
+    public static boolean inPartialCache(Path cache, Version version, DependencyHashCache depCache) throws IOException {
```
"inPartialCache"? I don't follow.

sciwhiz12 commented on src/main/java/net/neoforged/snowblower/tasks/BundlerExtractTask.java:
```diff
@@ -19,46 +21,124 @@
+
+            try (Stream<Path> walker = Files.walk(inRoot)) {
+                Iterable<Path> iterable = () -> walker.filter(Files::isRegularFile).iterator();
+                for (Path p : iterable) {
```
Technically not best practice (to wrap a `Stream`'s iterator into an `Iterable` given the contract of `Iterable`), but acceptable since this is guaranteed a one-off.

sciwhiz12 commented on README.md:
```diff
@@ -1,50 +1,95 @@
+* However, likely due to legal reasons, Mojang did not preserve the original local variable names in the released JARs (this was during the time when Minecraft's code was obfuscated).
+Instead, all local variables in every LVT had their name set to `☃` (`U+2603`), the Unicode codepoint for the snowman emoji.
+* This practice was maintained for *over 6 years* until 21w19a, when the identifier was changed to `â˜ƒ` for all local variables instead, thus marking the end of the snowman era.
+* In 21w37a, LVT identifiers were changed to instead be of the form `$$<index>`, counting up from `0` for every local variable defined in the table.
```
Additional tidbits:

21w19a is the first version that updated from Java 8, to Java 16 (then the latest version, I recall). That is probably linked with why the identifier changed in that version; potentially some sort of text encoding issue, since `â˜ƒ` looks like [mojibake](https://en.wikipedia.org/wiki/Mojibake) to me.

Java 16 included the feature of helpful `NullPointerException`s (see [JEP 358](https://openjdk.org/jeps/358), landed in Java 14), which now includes the name of the local variable in the exception message as stored in the LVT.

This is why some snapshots later (probably when Mojang got tired of seeing seemingly-garbled text in crash reports that had NPEs which may be confusing some people), Mojang changed to the auto-numbered names local variables instead.

sciwhiz12 commented on src/main/java/net/neoforged/snowblower/Generator.java:
```diff
@@ -254,30 +257,103 @@ public void run() throws IOException, GitAPIException {
+        // Allow resuming by finding the last thing we generated
+        int skipCount = this.getSkipCount(versions, filteredVersions, toGenerate, range[0]);
+        if (skipCount == -1) {
+            return;
```
I'd like to have a brief comment here to note that -1 means an error occured, so no further processing should be done. (`getSkipCount` already logs the error.)

sciwhiz12 reviewed (changes requested):
A read-through of it looks good. Mostly found some small details and some other comments.

We should update Gradle, the plugins (particularly to switch to `net.neoforged.licenser`), and the license header as it still says "Forge Development LLC" even for new files.

sciwhiz12 committed:
Simplify MinecraftVersion.Type

sciwhiz12 committed:
Add version pattern as comment

sciwhiz12 committed:
Replace isDev check with null check

Shut up an error from IDEA about a possible null dereference.

sciwhiz12 committed:
Remove unused parameters

sciwhiz12 committed:
Add support for April Fools 26w14a

sciwhiz12 committed:
Add a brief CONTRIBUTING.md for version ID bumping

sciwhiz12 committed:
Add comment explaining getSkipCount return value

sciwhiz12 committed:
Replace Collectors#toSet as modifiable set is needed

While Collectors#toSet doesn't specify a guarantee of modifiability, and
 it is unlikely that the implementation that returns a modifiable set
 will change anytime soon, we might as well fix it now that we've
 noticed it.

sciwhiz12 committed:
Update to Java 25

sciwhiz12 committed:
Remove --github-actions CLI parameter

The environment variable is sufficient to determine whether Snowblower
is running in a GitHub Actions runner.

sciwhiz12 committed:
Switch to NeoForged licenser plugin

sciwhiz12 committed:
Upgradle to 9.4.1

sciwhiz12 committed:
Update shadow plugin to 9.4.1

The shadow plugin changed ownership from johnrengleman to GradleUp.

sciwhiz12 committed:
Update GradleUtils to 5.1.1

sciwhiz12 committed:
Fix license header for new files

sciwhiz12 committed:
Resolve Gradle 9 warnings

sciwhiz12 committed:
Remove outdated reference in license configuration

sciwhiz12 reviewed (approved):
@SizableShrimp has assented to me finishing up this PR, so we can get this pulled into the main branch and further work be started as necessary.

sciwhiz12 committed:
Remove invalid trailing comma in JSON

sciwhiz12 committed:
Add workaround for detecting 26w14a as special

Mojang released an April Fools snapshot, 26w14a, matching the old
version format. This breaks the assumption that special versions, like
April Fools snapshots, have non-conventional versions that don't fit the
usual version formats (old or new; Mojang changed to the new format in
2026).

To workaround this, assume any snapshot versions that match the old
format but with the '26' year and beyond are assumed to be special
versions.

sciwhiz12 committed:
Use plural branch names and switch from 'dev' to 'snapshots'

The rename makes it clearer what the branch contains.
